sw/device/tests/dif/dif_rstmgr_unittest.cc

```diff
@@ -0,0 +1,306 @@
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "rstmgr_regs.h"  // Generated.
+
+namespace dif_rstmgr_unittest {
+namespace {
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Each;
+using testing::Eq;
+using testing::Test;
+
+class RstmgrTest : public Test, public MmioTest {
+ protected:
+  dif_rstmgr_t rstmgr_ = {
+      .params = {.base_addr = dev().region()},
+  };
+};
+
+class InitTest : public RstmgrTest {};
+
+TEST_F(InitTest, NullArgs) {
+  dif_rstmgr_params_t params = {.base_addr = dev().region()};
+  EXPECT_EQ(dif_rstmgr_init(params, nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(InitTest, Success) {
+  dif_rstmgr_t rstmgr;
+  dif_rstmgr_params_t params = {.base_addr = dev().region()};
+  EXPECT_EQ(dif_rstmgr_init(params, &rstmgr), kDifRstmgrOk);
+}
+
+class ResetTest : public RstmgrTest {};
+
+TEST_F(ResetTest, NullArgs) {
+  EXPECT_EQ(dif_rstmgr_reset(nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetTest, Success) {
+  EXPECT_WRITE32(RSTMGR_RESET_INFO_REG_OFFSET,
+                 std::numeric_limits<uint32_t>::max());
+  EXPECT_WRITE32(RSTMGR_SW_RST_CTRL_N_REG_OFFSET,
+                 std::numeric_limits<uint32_t>::max());
+
+  EXPECT_EQ(dif_rstmgr_reset(&rstmgr_), kDifRstmgrOk);
+}
+
+class ResetLockTest : public RstmgrTest {};
+
+TEST_F(ResetLockTest, NullArgs) {
+  EXPECT_EQ(dif_rstmgr_reset_lock(nullptr, 0), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetLockTest, BadPeripheral) {
+  EXPECT_EQ(dif_rstmgr_reset_lock(&rstmgr_, RSTMGR_PARAM_NUMSWRESETS),
+            kDifRstmgrBadArg);
+}
+
+TEST_F(ResetLockTest, Success) {
+  for (uint32_t bit_index = 0; bit_index < RSTMGR_PARAM_NUMSWRESETS;
+       ++bit_index) {
+    uint32_t bitfield = bitfield_bit32_write(
+        std::numeric_limits<uint32_t>::max(), bit_index, false);
+    EXPECT_WRITE32(RSTMGR_SW_RST_REGEN_REG_OFFSET, bitfield);
+
+    EXPECT_EQ(dif_rstmgr_reset_lock(&rstmgr_, bit_index), kDifRstmgrOk);
+  }
+}
+
+class ResetIsLockedTest : public RstmgrTest {};
+
+TEST_F(ResetIsLockedTest, NullArgs) {
+  bool is_locked;
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(nullptr, 0, nullptr), kDifRstmgrBadArg);
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(nullptr, 0, &is_locked),
+            kDifRstmgrBadArg);
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, 0, nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetIsLockedTest, BadPeripheral) {
+  bool is_locked;
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, RSTMGR_PARAM_NUMSWRESETS,
+                                       &is_locked),
+            kDifRstmgrBadArg);
+}
+
+TEST_F(ResetIsLockedTest, Success) {
+  for (uint32_t bit_index = 0; bit_index < RSTMGR_PARAM_NUMSWRESETS;
+       ++bit_index) {
+    uint32_t bit_locked = bitfield_bit32_write(
+        std::numeric_limits<uint32_t>::max(), bit_index, false);
+    EXPECT_READ32(RSTMGR_SW_RST_REGEN_REG_OFFSET, bit_locked);
```

Comment:
Maybe we should add some kind of mechanism for doing this that doesn't involve using bitfield.h? Not something we need to do in this PR, obviouslly.

Could you comment that this is supposed to be all ones except one bit? It wasn't obvious for me and I was going to incorrectly suggest you use EXPECT_READ's built-in syntax.

Reply:
For later: Do the other bits *have* to be all ones? Yes they will be the first time, but other software might have locked some resets and the DIF still has to work, so all the other bits shouldn't affect the outcome of this read.

Reply:
> Yes they will be the first time, but other software might have locked some resets and the DIF still has to work

This test simply checks whether this function correctly gets the information about the state of every peripheral. It doesn't cover any other use-cases, but the simplest way to make sure is to set all peripheral bits apart from one to high.